`CREATE` tests

OlivierBBB opened the pull request.

OlivierBBB committed:
first commit

OlivierBBB committed:
ras

OlivierBBB committed:
feat: wip tests

OlivierBBB committed:
feat: root level CREATE tests

OlivierBBB committed:
ras

OlivierBBB committed:
fix: added missing salt parameter to genericCreate()

OlivierBBB committed:
feat: some abort tests for CREATE(2)

OlivierBBB committed:
feat: abort / success and success / abort tests for CREATE(2)'s

OlivierBBB committed:
feat: more tests

OlivierBBB committed:
fix: documentation + no SUX for CREATE2 anymore

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
spotless

OlivierBBB committed:
fix: make the test work

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
spotless

OlivierBBB committed:
feat: added documentation for nontrivial CREATE tests

OlivierBBB committed:
feat: more CREATE test todo's

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
ras

OlivierBBB committed:
fix: CREATE's gasPaidOutOfPocket()

OlivierBBB committed:
fix: remove double use of CREATE's gasPaidOutOfPocket

OlivierBBB committed:
fix: using creator snapshots for creatorAccountFragment of CreateSection

OlivierBBB committed:
fix: CREATE2 gas pricing (similar error to CREATE pricing error)

fixes GAS_NEXT issues for now

OlivierBBB committed:
fix: CREATE2 address computations now trace the nonce

Not that the nonce is required, but it's part of the lookup

  HUB -> RLPADDR

OlivierBBB committed:
ras

OlivierBBB committed:
spotless

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java:
```diff
@@ -206,6 +206,7 @@ public CreateSection(Hub hub, MessageFrame frame) {
     if (failedCreate) {
       finalContextFragment = ContextFragment.nonExecutionProvidesEmptyReturnData(hub);
       scenarioFragment.setScenario(CREATE_FAILURE_CONDITION_WONT_REVERT);
+      commonValues.payGasPaidOutOfPocket(hub);
```
omission

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/section/CreateSection.java:
```diff
@@ -361,8 +364,8 @@ private void finishAbort(final Hub hub) {
-            preOpcodeCreateeSnapshot,
-            preOpcodeCreateeSnapshot,
+            preOpcodeCreatorSnapshot,
+            preOpcodeCreatorSnapshot,
```
typo

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/rlpaddr/RlpAddr.java:
```diff
@@ -109,7 +109,13 @@ public void callRlpAddrCreate2(MessageFrame frame, Bytes32 salt, Bytes32 hash) {
+            currentAddress,
+            salt,
+            hash,
+            longToUnsignedBigInteger(frame.getWorldUpdater().get(currentAddress).getNonce()));
```
Even though it's unnecessary, the nonce is part of the HUB -> RLPADDR lookup for CREATE2.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/rlpaddr/RlpAddr.java:
```diff
@@ -134,6 +142,7 @@ private void traceCreate2(int stamp, RlpAddrOperation chunk, Trace trace) {
           .saltLo(chunk.salt().orElseThrow().slice(LLARGE, LLARGE))
           .kecHi(chunk.keccak().orElseThrow().slice(0, LLARGE))
           .kecLo(chunk.keccak().orElseThrow().slice(LLARGE, LLARGE))
+          .nonce(nonceBytes)
```
We trace the nonce in all cases.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/rlpaddr/RlpAddrOperation.java:
```diff
@@ -49,8 +49,13 @@ public RlpAddrOperation(Bytes32 rawDepAddress, OpCode opCode, BigInteger nonce,
+      Address address,
+      Bytes32 salt,
+      Bytes32 kec,
+      BigInteger nonce) {
```
Adding the nonce to the interface

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/opcode/gas/projector/Create.java:
```diff
@@ -52,12 +52,13 @@ public long largestOffset() {
     }
+
+    long gasMinusUpfront = currentGas - upfrontGasCost;
+    return gasMinusUpfront - gasMinusUpfront / 64;
```
This was a genuine mistake. the 63/64-ths business happens after paying the upfront gas cost.

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/opcode/gas/projector/Create2.java:
```diff
@@ -57,13 +57,15 @@ public long linearPerWord() {
+
+    final long gasMinusUpfrontGasCost = currentGas - upfrontGasCost;
+
+    return gasMinusUpfrontGasCost - gasMinusUpfrontGasCost / 64;
```
Same mistake for CREATE2

OlivierBBB commented on testing/src/main/java/net/consensys/linea/testing/ToyAccount.java:
```diff
@@ -202,4 +202,9 @@ public ReferenceTestWorldState.AccountMock toAccountMock() {
+  public ToyAccount raiseNonceBy(long k) {
+    long updatedNonce = this.getNonce() + k;
+    return new ToyAccount(this.parent, this.getAddress(), updatedNonce, this.balance, this.balance);
+  }
```
Bandaid

OlivierBBB committed:
feat: upgraded the todo.md's to GitHub issues

OlivierBBB committed:
fix: RLPADDR needs to hash the nonce to disambiguate CREATE2's

We unconditionally import account.nonce() to the RLPADDR module. We must
therefore also include the nonce to disambiguate duplicate CREATE2's

lorenzogentile404 commented on arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/createTests/failure/Create.java:
```diff
@@ -0,0 +1,236 @@
+                  + 1 //
+              )
+          .op(JUMPI)
+          //
```
Could you add some comment about the purpose of the programs you are defining in this test class, especially for this long one above with this sum of 1s.

OlivierBBB replied:
Done.

lorenzogentile404 commented on arithmetization/src/test/java/net/consensys/linea/zktracer/instructionprocessing/createTests/trivial/RootLevel.java:
```diff
@@ -0,0 +1,193 @@
+      case THIRTY_TWO -> program.push(0x20);
+      case MSIZE -> program.op(MSIZE);
+      case MAX -> program.push("ff".repeat(32));
+    }
```
Here you may consider to do the following.

Modify the enum to:

```
public enum SizeParameter {
  ZERO(0),
  TWELVE(12), // - 3 - 1
  THIRTEEN(13), // - 3 + 0
  FOURTEEN(14), // - 3 + 1
  THIRTY_TWO(32),
  MSIZE,
  MAX;

 public int getValue() {
        if (this == MSIZE || this == MAX) {
            throw new UnsupportedOperationException("...");
        }
        return this.value;
    }

  public boolean isAnyOf(SizeParameter... sizeParameters) {
    for (SizeParameter sizeParameter : sizeParameters) {
      if (this == sizeParameter) {
        return true;
      }
    }
    return false;
  }

  public boolean willRaiseException() {
    return this.isAnyOf(MAX);
  }
}
```
Then, the code above would become:

```
 switch (sizeParameter) {
      case MSIZE -> program.op(MSIZE);
      case MAX -> program.push("ff".repeat(32));
      deafult -> program.push(sizeParameter.getValue();
    }
```
However, it is not a big deal to leave it as it is, but this could be a good approach for the future.

lorenzogentile404 replied:
Another approach may be using `Either`, where you basically either store an `int` or another type, that could be an `enum`, but it is probably not worth facing that complexity for this simple scenario.

lorenzogentile404 replied:
also, what translates to an integer may have a more meaningful name rather than `ZERO, TWELVE` etc.

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
fix: documentation for CreateInducedFailureTest

OlivierBBB committed:
spotless

OlivierBBB committed:
feat: Failure Condition F via nested CREATE2's

OlivierBBB committed:
fix: the OOB related call for CREATE's only requires NONCE and HAS_CODE of the createe if the createe gets loaded into the trace

OlivierBBB committed:
fix: minor test stuff

OlivierBBB committed:
spotless

OlivierBBB committed:
feat: a little less messy, but still very messy

OlivierBBB committed:
feat: moving stuff around ...

OlivierBBB committed:
fix: some wip refactoring on CreateSection

OlivierBBB committed:
ras: renaming for simpler interpretation of ParametricTests results

OlivierBBB committed:
fix: get latest deployment info for CREATE's that revert

down to 7 failing CREATE tests, none of which (AFAICT) related to the
CREATE's themselves.

OlivierBBB committed:
fix: make SELFDESTRUCT finish on a single context row

OlivierBBB committed:
fix: CallFrame's callValue (value) must be gleamed from the apparentValue

OlivierBBB committed:
fix: constraints point back to master

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
spotless

OlivierBBB committed:
fix: remove junit parallelism

OlivierBBB commented on arithmetization/src/main/java/net/consensys/linea/zktracer/module/hub/Hub.java:
```diff
@@ -625,7 +625,7 @@ public void traceContextEnter(MessageFrame frame) {
           newChildContextNumber(),
           this.deploymentStatusOf(frame.getContractAddress()),
-          frame.getValue(),
+          frame.getApparentValue(),
```
Important bug fix, required e.g. for DELEGATECALL to provide the CallFrame with the correct `value` i.e. `callValue`

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
fix: re-did the type change for scenario/XXX_GAS columns

Otherwise the project doesn't work properly with updated constraints

OlivierBBB committed:
feat: new simplified tests

triggers

  scenario/CALL_FAILURE_WILL_REVERT

using a self call and only 2 or 3 iterations (and blows up for
STATICCALL) with an exception.

OlivierBBB committed:
Merge branch 'arith-dev' into CREATE-tests

OlivierBBB committed:
spotless
